### PR TITLE
Some more citations and fixes to complexities tip

### PR DESCRIPTION
**Pull request by Benjamin-Lee**
**Did you add yourself as a [contributor](https://github.com/Benjamin-Lee/deep-rules/blob/master/contributors.md) if this is your first contribution?**

- [x] Yes, I added myself or am already a contributor

**Any more details?**

### CODE REVIEW

File: content/05.dl-complexities.md [hunk 1]
```diff
@@ -2,13 +2,13 @@
 
 Correctly training deep neural networks is a non-trivial process.
 There are many different options and potential pitfalls at every stage.
-To get good results you have to expect to train many networks with a range of different parameter and hyperparameter settings.
+To get good results, you must expect to train many networks with a range of different parameter and hyperparameter settings.
 Despite improved framework ease-of-use and on-demand cloud computing resources this means DL can be very demanding, often requiring significant infrastructure and patience to achieve state-of-the-art performance [@doi:10.1109/JPROC.2017.2761740].
```
Comment (Benjamin-Lee) [inline content/05.dl-complexities.md]: This line doesn't seem to correct, but I'm not quite sure what framework ease-of-use and DL infrastructure requirements have in common. Can we make this line clearer?

File: content/05.dl-complexities.md [hunk 2]
```diff
@@ -2,13 +2,13 @@
...
 The experimentation inherent to DL is often noisy (requiring repetition) and represents a significant organizational challenge.
-All code, random seeds, parameters, and results must be carefully corralled using good coding practices (for example, version control, continuous integration etc.) in order to be effective and interpretable.
-This organization is also key to being able to efficiently share your work and to update your model as new data becomes available. 
+All code, random seeds, parameters, and results must be carefully corralled using general good coding practices (for example, version control [@doi:10.1371/journal.pcbi.1004947], continuous integration etc.) in order to be effective and interpretable.
+This organization is also key to being able to efficiently share and reproduce your work as well as to update your model as new data becomes available [@doi:10.1371/journal.pcbi.1003285; @arxiv:1810.08055].
```
Comment (agitter) [inline content/05.dl-complexities.md]: Perhaps move these citations to follow `reproduce your work`?  They have more to do with that than updating models.

File: content/05.dl-complexities.md (hunk 1 above)
Comment (Benjamin-Lee) [reply content/05.dl-complexities.md]: To make this syntactically (but not necessarily semantically) correct, how about:

"Despite improved framework ease-of-use and on-demand cloud computing resources, DL can be very demanding, often requiring significant infrastructure and patience to achieve state-of-the-art performance [@doi:10.1109/JPROC.2017.2761740]."

Comment (agitter) [reply content/05.dl-complexities.md]: Should we start the sentence with "DL can be very demanding" and move the initial clause elsewhere?  I think the original idea was that even though it is easier to implement and run DL models, that doesn't change the local or cloud computational demands.

Comment (Benjamin-Lee) [reply content/05.dl-complexities.md]: How about:
> DL [training?] can be very demanding, often requiring significant computing infrastructure and patience to achieve state-of-the-art performance [@doi:10.1109/JPROC.2017.2761740].

Comment (agitter) [reply content/05.dl-complexities.md]: Yes, I like that.  I lean toward excluding "training" in case that implies training a single model is hard.  Often the real demand is in hyperparameter search and training many models.

Comment (Benjamin-Lee) [reply content/05.dl-complexities.md]: What about "model optimization" or "hyperparameter search" instead of training?

Comment (Benjamin-Lee) [reply content/05.dl-complexities.md]: You're right; there's no need for that word since the last sentence makes it clear what we're talking about

File: content/05.dl-complexities.md [hunk 3]
```diff
@@ -2,13 +2,13 @@
 
 Correctly training deep neural networks is a non-trivial process.
 There are many different options and potential pitfalls at every stage.
-To get good results you have to expect to train many networks with a range of different parameter and hyperparameter settings.
-Despite improved framework ease-of-use and on-demand cloud computing resources this means DL can be very demanding, often requiring significant infrastructure and patience to achieve state-of-the-art performance [@doi:10.1109/JPROC.2017.2761740].
+To get good results, you must expect to train many networks with a range of different parameter and hyperparameter settings.
+Deep learning can be very demanding, often requiring significant computing infrastructure and patience to achieve state-of-the-art performance [@doi:10.1109/JPROC.2017.2761740].
```
Comment (rasbt) [inline content/05.dl-complexities.md]: > significant computing infrastructure

maybe "extensive" instead of "significant?"

File: content/05.dl-complexities.md [hunk 4]
```diff
@@ -2,13 +2,13 @@
...
-All code, random seeds, parameters, and results must be carefully corralled using good coding practices (for example, version control, continuous integration etc.) in order to be effective and interpretable.
-This organization is also key to being able to efficiently share your work and to update your model as new data becomes available. 
+All code, random seeds, parameters, and results must be carefully corralled using general good coding practices (for example, version control [@doi:10.1371/journal.pcbi.1004947], continuous integration etc.) in order to be effective and interpretable.
+This organization is also key to being able to efficiently share and reproduce your work [@doi:10.1371/journal.pcbi.1003285; @arxiv:1810.08055] as well as to update your model as new data becomes available.
 
-One specific reproducibility pitfall that is often missed is the default use of non-deterministic algorithms by CUDA/CuDNN backends when training on GPUs.
+One specific reproducibility pitfall that is often missed in deep learning applications is the default use of non-deterministic algorithms by CUDA/CuDNN backends when training models on GPUs.
```
Comment (rasbt) [inline content/05.dl-complexities.md]: technically, these algorithms are also non-deterministic during prediction (after training). It's just not as noticeable. I would rephrase to 

> ... by CUDA/CuDNN backends when ~~training models on~~ [using] GPUs.

File: content/05.dl-complexities.md [hunk 5]
```diff
@@ -2,13 +2,13 @@
...
 Making this process reproducible is distinct from setting random seeds, which will primarily affect pseudorandom deterministic procedures such as shuffling and initialization, and requires explicitly specifying the use of deterministic algorithms in your DL library [@url:https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#reproducibility]. 
 
 Similar to [Tip 4](#baselines), try to start with a relatively smaller network and increase the size and complexity as needed to prevent wasting time and resources.
```
Comment (rasbt) [inline content/05.dl-complexities.md]: > ...relatively ~~smaller~~ [small] network a ...

**Comment by Benjamin-Lee** [comment]
@rasbt Done!
